snake-dqn/snake_game.js

```diff
@@ -0,0 +1,332 @@
+    assertPositiveInteger(args.initLen, 'initLen');
+
+    this.height_ = args.height;
+    this.width_ = args.width;
+    this.numFruits_ = args.numFruits;
+    this.initLen_ = args.initLen;
+
+    this.reset();
+  }
+
+  /**
+   * Reset the state of the game.
+   *
+   * @return {object} Initial state of the game.
+   *   See the documentation of `getState()` for details.
+   */
+  reset() {
+    this.initializeSnake_();
+    this.makeFruits_();
+    return this.getState();
+  }
+
+  /**
+   * Perform a step of the game.
+   *
+   * @param {0 | 1 | 2 | 3} action The action to take in the current step.
+   *   The meaning of the possible values:
+   *     0 - left
+   *     1 - top
+   *     2 - right
+   *     3 - bottom
+   * @return {object} Object with the following keys:
+   *   - `reward` {number} the reward value.
+   *     - 0 if no fruit is eaten in this step
+   *     - 1 if a fruit is eaten in this step
+   *   - `state` New state of the game after the step.
+   *   - `done` {boolean} whether the game has ended after this step.
+   *     A game ends when the head of the snake goes off the board or goes
+   *     over its own body.
+   */
+  step(action) {
+    const [headY, headX] = this.snakeSquares_[0];
+
+    // Calculate the coordinates of the new head and check whether it has
+    // gone off the board, in which case the game will end.
+    let done;
+    let newHeadY;
+    let newHeadX;
+    if (action === ACTION_LEFT) {
+      newHeadY = headY;
+      newHeadX = headX - 1;
+      done = newHeadX < 0;
+    } else if (action === ACTION_UP) {
+      newHeadY = headY - 1;
+      newHeadX = headX;
+      done = newHeadY < 0
+    } else if (action === ACTION_RIGHT) {
+      newHeadY = headY;
+      newHeadX = headX + 1;
+      done = newHeadX >= this.width_;
+    } else if (action === ACTION_DOWN) {
+      newHeadY = headY + 1;
+      newHeadX = headX;
+      done = newHeadY >= this.height_;
+    } else {
+      throw new Error(
+          `Invalid action: ${action}. Valid actions are: ` +
+          `${ACTOIN_LEFT}, ${ACTION_UP}, ${ACTION_RIGHT}, ${ACTION_DOWN}`);
+    }
+
+    // Check if the head goes over the snake's body, in which case the
+    // game will end.
+    for (let i = 1; i < this.snakeSquares_.length; ++i) {
+      if (this.snakeSquares_[i][0] === newHeadY &&
+          this.snakeSquares_[i][1] === newHeadX) {
+        done = true;
+      }
+    }
+
+    if (done) {
+      return {reward: DEATH_REWARD, done};
+    }
+
+    // Update the position of the snake.
+    this.snakeSquares_.pop();
+    this.snakeSquares_.unshift([newHeadY, newHeadX]);
+
+    // Check if a fruit is eaten.
+    let reward = NO_FRUIT_REWARD;
+    for (let i = 0; i < this.fruitSquares_.length; ++i) {
+      const fruitYX = this.fruitSquares_[i];
+      if (fruitYX[0] === newHeadY && fruitYX[1] === newHeadX) {
+        reward = FRUIT_REWARD;
+        this.fruitSquares_.splice(i, 1);
+        this.makeFruits_();
```

Comment:
Usually the snake grows if it eats a fruit.  It looks like you're not using that in this variant?